File: problem-solving/lc-474-medium-ones-and-zeroes/2-bottom-up.py
"""
https://leetcode.com/problems/ones-and-zeroes/

Complexity
----------
Time: O(L*m*n)  // L = number of strings
Space: O(m*n)
"""

class Solution:
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        for s in strs:
            ones = 0
            for c in s:
                if c == '1':
                    ones += 1
            zeros = len(s) - ones
            for z in range(m, zeros - 1, -1):
                for o in range(n, ones - 1, -1):
                    dp[z][o] = max(dp[z][o], dp[z - zeros][o - ones] + 1)
        return dp[m][n]

File: problem-solving/lc-516-medium-longest-palindromic-subsequence/2-bottom-up.go
/*
https://leetcode.com/problems/longest-palindromic-subsequence/

Complexity
----------
Time: O(n^2)
Space: O(n^2)

Another approach
----------------
Reverse given string and find LCS b/w given string and reversed: https://leetcode.com/problems/longest-palindromic-subsequence/solutions/99151/super-simple-solution-using-reversed-string/
*/

func longestPalindromeSubseq(s string) int {
    n := len(s)
    var dp [][]int

    for i := 0; i < n; i++ {
        dp = append(dp, make([]int, n))
    }

    for i:= n - 1; i >=0; i-- {
        dp[i][i] = 1
        for j := i + 1; j < n; j++ {
            if s[i] == s[j] {
                dp[i][j] = 2 + dp[i + 1][j - 1]
            } else {
                dp[i][j] = max(dp[i + 1][j], dp[i][j - 1])
            }
        }
    }
    return dp[0][n - 1]
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}
